### PR TITLE
bpo-46933: Make pwd module optional (GH-31700)

### PR DESCRIPTION
**Pull request by tiran**
https://bugs.python.org/issue46933

### CODE REVIEW

**Comment by erlend-aasland** [comment]
Hm, for some reason, `configure` now decides _not_ to build `pwd` on my Mac, thus breaking `posixpath.expanduser` for my username:

```bash
$ python3.11
Python 3.11.0a5 (v3.11.0a5:c4e4b91557, Feb  3 2022, 14:54:01) [Clang 13.0.0 (clang-1300.0.29.30)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import posixpath
>>> posixpath.expanduser("~erlendaasland")
'/Users/erlendaasland'
>>> ^D
$ ./python.exe                 
Python 3.11.0a5+ (heads/no-pwd-dirty:a6ca976941, Mar  6 2022, 00:31:10) [Clang 13.0.0 (clang-1300.0.29.30)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import posixpath
>>> posixpath.expanduser("~erlendaasland")
'~erlendaasland'
>>> ^D
```

`getpwuid` and `getpwuid_r` are both available, so I don't see why.

```bash
$ grep getpwuid config.cache
ac_cv_func_getpwuid=${ac_cv_func_getpwuid=yes}
ac_cv_func_getpwuid_r=${ac_cv_func_getpwuid_r=yes}
```

**Comment by erlend-aasland** [comment]
Aaaah, I see it!

https://github.com/python/cpython/blob/6927632492cbad86a250aa006c1847e03b03e70b/configure.ac#L6465-L6466

https://github.com/python/cpython/blob/6927632492cbad86a250aa006c1847e03b03e70b/configure.ac#L6378

**Review by erlend-aasland — CHANGES_REQUESTED** [review]
We need to use `m4_foreach_w` (or `m4_map_args_w`) in `PY_STDLIB_MOD` and `PY_STDLIB_MOD_SIMPLE` when checking `py_stdlib_not_available` for unavailable modules. There may be a more elegant solution.

**Comment by tiran** [comment]
> We need to use `m4_foreach_w` (or `m4_map_args_w`) in `PY_STDLIB_MOD` and `PY_STDLIB_MOD_SIMPLE` when checking `py_stdlib_not_available` for unavailable modules. There may be a more elegant solution.

I tried ``m4_set_add_all`` but it didn't work like I hoped. The m4 set functions seem to get evaluated during autoconf. We need to evaluate them when ``configure`` is running.

**Comment by erlend-aasland** [comment]
> The m4 set functions seem to get evaluated during autoconf. We need to evaluate them when configure is running.

Yeah, I tried some experiments with the m4 wrappers yesterday, but they didn't work out.

**Comment by tiran** [comment]
> Yeah, I tried some experiments with the m4 wrappers yesterday, but they didn't work out.

We may have to restructure the test. I have a working solution, will clean it up and push it later today.

File: configure.ac [hunk 1]
```diff
@@ -6450,10 +6460,10 @@ MODULE_BLOCK=
 
 dnl Check for stdlib extension modules
 dnl PY_STDLIB_MOD([NAME], [ENABLED-TEST], [SUPPORTED-TEST], [CFLAGS], [LDFLAGS])
-dnl sets MODULE_$NAME based on $py_stdlib_not_available, ENABLED-TEST,
+dnl sets MODULE_$NAME based on $PY_STDLIB_MOD_SET_NA, ENABLED-TEST,
```
Comment (erlend-aasland) [inline configure.ac]: Comment nit. Something like this, or reword :)

```suggestion
dnl sets MODULE_$NAME based on PY_STDLIB_MOD_SET_NA(), ENABLED-TEST,
```

File: configure.ac [hunk 2]
```diff
@@ -6450,10 +6460,10 @@ MODULE_BLOCK=
...
 dnl and SUPPORTED_TEST. ENABLED-TEST and SUPPORTED-TEST default to true if
 dnl empty.
-dnl    n/a: $NAME in $py_stdlib_not_available (not available on platform)
+dnl    n/a: $NAME in $PY_STDLIB_MOD_SET_NA (not available on platform)
```
Comment (erlend-aasland) [inline configure.ac]: More comment nit.

```suggestion
dnl    n/a: marked unavailable on platform by PY_STDLIB_MOD_SET_NA()
```

File: configure.ac [hunk 3]
```diff
@@ -6480,16 +6490,14 @@ AC_DEFUN([PY_STDLIB_MOD], [
 ])
 
 dnl Define simple stdlib extension module
-dnl Always enable unless the module is listed in py_stdlib_not_available
+dnl Always enable unless the module is listed in PY_STDLIB_MOD_SET_NA
```
Comment (erlend-aasland) [inline configure.ac]:
```suggestion
dnl Always enable unless the module is disabled by PY_STDLIB_MOD_SET_NA
```

**Review by erlend-aasland — APPROVED** [review]
Yep, this approach works swell. Tested on my mac; all good. I left some comments nits; do with them as you please :)

**Pull request merged**